app/services/partnerships/report.rb

```diff
@@ -0,0 +1,78 @@
+# frozen_string_literal: true
+
+module Partnerships
+  class Report
+    CHALLENGE_WINDOW = 14.days.freeze
+    REMINDER_EMAIL_DELAY = 7.days.freeze
+
+    def self.call(**args)
+      new(**args).call
+    end
+
+    def initialize(cohort_id:, school_id:, lead_provider_id:, delivery_partner_id:)
+      @cohort_id = cohort_id
+      @school_id = school_id
+      @lead_provider_id = lead_provider_id
+      @delivery_partner_id = delivery_partner_id
+    end
+
+    def call
+      Partnership.transaction do
+        partnership = Partnership.find_or_initialize_by(
+          school_id: school_id,
+          cohort_id: cohort_id,
+          lead_provider_id: lead_provider_id,
+        )
+
+        partnership.challenge_reason = partnership.challenged_at = nil
+        partnership.delivery_partner_id = delivery_partner_id
+        partnership.pending = school_cohort.core_induction_programme?
+        partnership.challenge_deadline = CHALLENGE_WINDOW.from_now
+        partnership.report_id = SecureRandom.uuid
+        partnership.save!
+
+        partnership.event_logs.create!(
+          event: :reported,
+        )
+
+        PartnershipNotificationService.new.delay.notify(partnership)
+        PartnershipReminderJob.set(wait: REMINDER_EMAIL_DELAY).perform_later(
+          partnership: partnership,
+          report_id: partnership.report_id,
+        )
+
+        if partnership.pending?
+          PartnershipActivationJob.new.delay(run_at: partnership.challenge_deadline).perform(
+            partnership: partnership,
+            report_id: partnership.report_id,
+          )
+        end
+
+        partnership
```

Comment:
Do we need to return it for anything other than testing? If not, can we retrieve it differently in testing? Say, a database call fetching a partnership between a school and a lead provider?

Reply:
I think the question is really "what would you expect a service called `Partnerhsip::Report` return" rather than `where is the return value being used`. This service should have no knowledge of how and when it is being used and instead it should try to be as useful as possible. As it's only job is to create a new partnership, returning that partnership feels as a natural result of the execution (however I'd much rather prefer to have Success/Failure monad returned with some associated data).